// posthog-js/lib/src/heatmaps.js
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
import { includes, registerEvent } from './utils';
import RageClick from './extensions/rageclick';
import { document, window } from './utils/globals';
import { getParentElement, isTag } from './autocapture-utils';
function elementOrParentPositionMatches(el, matches, breakOnElement) {
    var curEl = el;
    while (curEl && !isTag(curEl, 'body')) {
        if (curEl === breakOnElement) {
            return false;
        }
        if (includes(matches, window === null || window === void 0 ? void 0 : window.getComputedStyle(curEl).position)) {
            return true;
        }
        curEl = getParentElement(curEl);
    }
    return false;
}
var Heatmaps = /** @class */ (function () {
    function Heatmaps(instance) {
        this.rageclicks = new RageClick();
        this._isDisabledServerSide = null;
        this._initialized = false;
        this.instance = instance;
    }
    Heatmaps.prototype.startIfEnabled = function () {
        if (this.isEnabled && !this._initialized) {
            this._setupListeners();
        }
    };
    Object.defineProperty(Heatmaps.prototype, "isEnabled", {
        get: function () {
            return !!this.instance.config.__preview_heatmaps;
        },
        enumerable: false,
        configurable: true
    });
    Heatmaps.prototype.getAndClearBuffer = function () {
        var buffer = this.buffer;
        this.buffer = undefined;
        return buffer;
    };
    Heatmaps.prototype._setupListeners = function () {
        var _this = this;
        if (!window || !document) {
            return;
        }
        registerEvent(document, 'click', function (e) { return _this._onClick((e || (window === null || window === void 0 ? void 0 : window.event))); }, false, true);
        registerEvent(document, 'mousemove', function (e) { return _this._onMouseMove((e || (window === null || window === void 0 ? void 0 : window.event))); }, false, true);
        this._initialized = true;
    };
    Heatmaps.prototype._getProperties = function (e, type) {
        // We need to know if the target element is fixed or not
        // If fixed then we won't account for scrolling
        // If not then we will account for scrolling
        var scrollY = this.instance.scrollManager.scrollY();
        var scrollX = this.instance.scrollManager.scrollX();
        var scrollElement = this.instance.scrollManager.scrollElement();
        var isFixedOrSticky = elementOrParentPositionMatches(e.target, ['fixed', 'sticky'], scrollElement);
        return {
            x: e.clientX + (isFixedOrSticky ? 0 : scrollX),
            y: e.clientY + (isFixedOrSticky ? 0 : scrollY),
            target_fixed: isFixedOrSticky,
            type: type,
        };
    };
    Heatmaps.prototype._onClick = function (e) {
        var _a;
        var properties = this._getProperties(e, 'click');
        if ((_a = this.rageclicks) === null || _a === void 0 ? void 0 : _a.isRageClick(e.clientX, e.clientY, new Date().getTime())) {
            this._capture(__assign(__assign({}, properties), { type: 'rageclick' }));
        }
        // TODO: Detect deadclicks
        this._capture(properties);
    };
    Heatmaps.prototype._onMouseMove = function (e) {
        var _this = this;
        clearTimeout(this._mouseMoveTimeout);
        this._mouseMoveTimeout = setTimeout(function () {
            _this._capture(_this._getProperties(e, 'mousemove'));
        }, 500);
    };
    Heatmaps.prototype._capture = function (properties) {
        if (!window) {
            return;
        }
        var url = window.location.href;
        this.buffer = this.buffer || {};
        if (!this.buffer[url]) {
            this.buffer[url] = [];
        }
        this.buffer[url].push(properties);
    };
    return Heatmaps;
}());
export { Heatmaps };
